File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>
#include <iostream>
using namespace std;

map<string, int> m;
vector<int> maxCount (11, 0);
vector<bool> check(11, false);

void Make(string& s, int idx, string course) {
    if (check[course.length()]) {
        ++m[course];
        
        if (maxCount[course.length()] < m[course])
            maxCount[course.length()] = m[course];
    }
    
    if (idx >= s.length()) {
        return;
    }
    
    for(int i = idx; i < s.length(); ++i) {
        Make(s, i + 1, course + s[i]);
    }
}

void MakeComb(string s) {
    Make(s, 0, "");
}


vector<string> solution(vector<string> orders, vector<int> course) {
    vector<string> answer;

    
    for(int i = 0; i < course.size(); ++i) {
        check[course[i]] = true;
    }
    
    for(int i = 0; i < orders.size(); ++i) {
        string menu = orders[i];
        sort(menu.begin(), menu.end());
        MakeComb(menu);
    }
    
    for(const auto item : m) {
        string menu = item.first;
        int count = item.second;
        
        if (count == maxCount[menu.length()] && count >= 2) {
            answer.push_back(menu);
        }
    }
    
    sort(answer.begin(), answer.end());
    return answer;
}

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public static Map<String, Integer> courseMap = new HashMap<>();

    public static class Course {
        String course;
        int count;

        public Course(String course, int count) {
            this.course = course;
            this.count = count;
        }
    }

    public String[] solution(String[] orders, int[] course) {
        Map<Integer, List<Course>> courses = new HashMap<>();
        Map<Integer, Integer> maxCourseSize = new HashMap<>();
        for (int i = 0; i < course.length; i++) {
            List<Course> courseList = new ArrayList<>();
            maxCourseSize.put(course[i], 0);
            courses.put(course[i], courseList);
        }

        for (String order : orders) {
            char[] charArray = order.toCharArray();
            Arrays.sort(charArray);
            for (int i = 0; i < course.length; i++) {
                if (course[i] > order.length()) continue;
                generate(0, course[i], 0, new StringBuilder(), charArray);
            }
        }

        for (String key : courseMap.keySet()) {
            int courseSize = key.length();
            int occurrences = courseMap.get(key);

            List<Course> courseList = courses.get(courseSize);
            courseList.add(new Course(key, occurrences));
            courses.put(courseSize, courseList);

            if (maxCourseSize.get(courseSize) < occurrences && occurrences >= 2) {
                maxCourseSize.put(courseSize, occurrences);
            }
        }

        List<String> result = new ArrayList<>();
        for (int courseSize : courses.keySet()) {
            List<Course> courseList = courses.get(courseSize);
            int maxCourse = maxCourseSize.get(courseSize);
            for (Course c : courseList) {
                if (c.count == maxCourse) {
                    result.add(c.course);
                }
            }
        }
        return result.stream()
                .sorted()
                .toArray(String[]::new);
    }

    public void generate(int curSize, int targetSize, int idx, StringBuilder temp, char[] charArray) {
        if (curSize == targetSize) {

            courseMap.put(temp.toString(), courseMap.getOrDefault(temp.toString(), 0) + 1);
            return;
        }

        for (int i = idx; i < charArray.length; i++) {
            temp.append(charArray[i]);
            generate(curSize + 1, targetSize, i + 1, new StringBuilder(temp), charArray);
            temp.deleteCharAt(temp.length() - 1);
        }
    }
}   